bindings/c/include/opendal.h

```diff
@@ -25,14 +25,229 @@
+
+ The [`OperatorPtr`] also has a transparent layout, allowing you
+ to check its validity by native boolean operator.
+ e.g. you could check by (!ptr) on a opendal_operator_ptr type
+ */
+typedef const void *opendal_operator_ptr;
+
+/*
+ The error type returned of all opendal APIs in the C code
+ */
+typedef struct opendal_error {
+  /*
+   The error number of the actual error
+   */
+  enum opendal_errno errno;
+  /*
+   The error status of the error, please see opendal_error_status
+   for more details
+   */
+  enum opendal_error_status error_status;
+  /*
+   The error_msg consumes an area of heap memory, please call
+   [`opendal_free_error`] on the error to free the memory
+   */
+  const char *error_msg;
+} opendal_error;
+
+/*
+ The Rust-like Result type of opendal C binding, it contains
+ a opendal_operator_ptr and a pointer to an error type.
+ Note that if `error` is not null, the error consumes heap memory
+ remember to free it by using [`opendal_free_error`]
+ */
+typedef struct opendal_result_op {
+  opendal_operator_ptr op_ptr;
+  struct opendal_error *error;
+} opendal_result_op;
+
+/*
+ The Rust-like Result type of opendal C binding, it contains
+ whether the write operation succeeds and a pointer to an error type.
+ Note that if `error` is not null, the error consumes heap memory
+ remember to free it by using [`opendal_free_error`]
+ */
+typedef struct opendal_result_write {
+  bool success;
+  struct opendal_error *error;
+} opendal_result_write;
+
+/*
+ The [`Bytes`] type is a C-compatible substitute for [`Bytes`]
+ in Rust, it will not be deallocated automatically like what
+ has been done in Rust. Instead, you have to call [`opendal_free_bytes`]
+ to free the heap memory to avoid memory leak.
+ The field `data` should not be modified since it might causes
+ the reallocation of the Vector.
+ */
+typedef struct opendal_bytes {
+  const uint8_t *data;
+  uintptr_t len;
+} opendal_bytes;
+
+/*
+ The Rust-like Result type of opendal C binding, it contains
+ the data that the read operation returns and a pointer to an error type.
+ If the read operation failed, the `data` fields should be a nullptr.
+ Note that if `error` is not null, the error consumes heap memory
+ remember to free it by using [`opendal_free_error`]
+ */
+typedef struct opendal_result_read {
+  struct opendal_bytes *data;
+  struct opendal_error *error;
+} opendal_result_read;
+
 #ifdef __cplusplus
 extern "C" {
 #endif // __cplusplus
 
 /*
- Hello, OpenDAL!
+ Returns a result type [`opendal_result_op`], with operator_ptr. If the construction succeeds
+ the error is nullptr, otherwise it contains the error information.
+
+ # Safety
+
+ It is [safe] under two cases below
+ * The memory pointed to by `scheme` must contain a valid nul terminator at the end of
+   the string.
+ * The `scheme` points to NULL, this function simply returns you a null opendal_operator_ptr
+ */
+struct opendal_result_op opendal_new_operator(const char *scheme);
```

Comment:
I don't like the idea of adding `result` for every operation.How about providing API like this:

```c
struct opendal_operator *opendal_operator_new(const char *scheme);
struct opendal_code *opendal_operator_build(struct opendal_operator *op);
```

Maybe we can add `opendal_code` like `hyper` does?

Reply:
The `hyper_code` is actually a concept like the `errno` I used.

I don't think this `two-stage` API is common in C.

Using API with only code or with extra error message is actually a trade-off IMO. Error number is commonly used in C and error message in commonly used in Rust.